octodns/provider/dyn.py

```diff
@@ -841,6 +1101,223 @@ def _mod_geo_Delete(self, dyn_zone, change):
+        pool = self._find_or_create_dynamic_pool(td, pools, 'default',
+                                                 new._type, values,
+                                                 record_extras={
+                                                     'automation': 'manual',
+                                                     'eligible': True,
+                                                 })
+        # There's no way in the client lib to create a ruleset with an existing
+        # pool (ref'd by id) so we have to do this round-a-bout.
+        active_pools = {
+            # TODO: disallow default as a pool id
+            'default': pool.response_pool_id
+        }
+        ruleset.add_response_pool(pool.response_pool_id)
+
+        # Get our monitor
+        monitor_id = self._traffic_director_monitor(new).dsf_monitor_id
+
+        # Make sure we have all the pools we're going to need
+        for _id, pool in sorted(new.dynamic.pools.items()):
+            values = [{
+                'weight': v.get('weight', 1),
+                'value': v['value'],
+            } for v in pool.data['values']]
+            pool = self._find_or_create_dynamic_pool(td, pools, _id,
+                                                     new._type, values,
+                                                     monitor_id)
+            active_pools[_id] = pool.response_pool_id
+
+        # Run through and configure our rules
+        for rule_num, rule in enumerate(reversed(new.dynamic.rules)):
+            criteria = defaultdict(lambda: defaultdict(list))
+            criteria_type = 'always'
+            try:
+                geos = rule.data['geos']
+                criteria_type = 'geoip'
+            except KeyError:
+                geos = []
+
+            for geo in geos:
+                geo = GeoCodes.parse(geo)
+                if geo['province_code']:
+                    criteria['geoip']['province'] \
+                        .append(geo['province_code'].lower())
+                elif geo['country_code']:
+                    criteria['geoip']['country'] \
+                        .append(geo['country_code'])
+                else:
+                    criteria['geoip']['region'] \
+                        .append(self.REGION_CODES[geo['continent_code']])
+
+            label = '{}:{}'.format(rule_num, uuid4().hex)
+            ruleset = DSFRuleset(label, criteria_type, [], criteria)
+            # Something you have to call create others the constructor does it
+            ruleset.create(td, index=insert_at)
+
+            # Add the primary pool for this rule
+            rule_pool = rule.data['pool']
+            ruleset.add_response_pool(active_pools[rule_pool])
+
+            # OK, we have the rule and its primary pool setup, now look to see
+            # if there's a fallback chain that needs to be configured
+            fallback = new.dynamic.pools[rule_pool].data.get('fallback', None)
+            while fallback:
```

Comment:
Heh. I started out trying to handle that initially and decided to simplify it to avoid getting hung up. Meant to TODO it, but guess I forgot. Will update shortly.

Reply:
Actually looks like I added loop detection over in https://github.com/github/octodns/blob/7290f14278eaecabf4b8f7bc88fac1980784096f/octodns/record/__init__.py#L510-L522 and there are tests for it over in https://github.com/github/octodns/blob/master/tests/test_octodns_record.py#L2501

That should mean that things aren't able to make it to this point with loops in them. I can go ahead and add simple loop detection here since it won't hurt anything.